# Multiple Linear Regression using Python
In this notebook we will solve a Multiple Linear Regression problem using Python programming language. In order to do this we will use some commonly used libraries for solving machine learning tasks, these includes Pandas, Numpy, Matplotlib and Scikit-learn.

## Task 1:

Suppose we want to determine the price per unit area of a house in a dataset of historical purchases made in an area in which we have multiple features such as __*transaction_number, transaction_date, house_age, distance_to_nearest_bus_station, number_of_convenience_stores, latitude, and longitude.*__


The objective of this task is to train a ML regression model that can accurately predict the price per unit area of a given house.

![alt text](img/real_estate.jpg "Real Estate")

### Step 1: Importing Libraries for our task

In [80]:
import pandas as pd
from matplotlib import pyplot as plt
import numpy as np

from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error, r2_score

### 2. Reading the data from the provided file
Since the data is in a CSV format, we are going to read this file using the pandas's read_csv method

In [81]:
df = pd.read_csv('dataset/03-real_estate_dataset.csv')

### 3. Exploring the data 
Let us explore the data 

In [82]:
df.shape

(414, 8)

In [83]:
df.head()

number  transaction_date  house_age  distance_to_nearest_bus_station  \
0       1          2012.917       32.0                         84.87882   
1       2          2012.917       19.5                        306.59470   
2       3          2013.583       13.3                        561.98450   
3       4          2013.500       13.3                        561.98450   
4       5          2012.833        5.0                        390.56840   

   number_of_convenience_stores  latitude  longitude  unit_area_price  
0                            10  24.98298  121.54024             37.9  
1                             9  24.98034  121.53951             42.2  
2                             5  24.98746  121.54391             47.3  
3                             5  24.98746  121.54391             54.8  
4                             5  24.97937  121.54245             43.1

In [84]:
df.describe()

number  transaction_date   house_age  \
count  414.000000        414.000000  414.000000   
mean   207.500000       2013.148971   17.712560   
std    119.655756          0.281967   11.392485   
min      1.000000       2012.667000    0.000000   
25%    104.250000       2012.917000    9.025000   
50%    207.500000       2013.167000   16.100000   
75%    310.750000       2013.417000   28.150000   
max    414.000000       2013.583000   43.800000   

       distance_to_nearest_bus_station  number_of_convenience_stores  \
count                       414.000000                    414.000000   
mean                       1083.885689                      4.094203   
std                        1262.109595                      2.945562   
min                          23.382840                      0.000000   
25%                         289.324800                      1.000000   
50%                         492.231300                      4.000000   
75%                        1454.279000                      6.000000   
max                        6488.021000                     10.000000   

         latitude   longitude  unit_area_price  
count  414.000000  414.000000       414.000000  
mean    24.969030  121.533361        37.980193  
std      0.012410    0.015347        13.606488  
min     24.932070  121.473530         7.600000  
25%     24.963000  121.528085        27.700000  
50%     24.971100  121.538630        38.450000  
75%     24.977455  121.543305        46.600000  
max     25.014590  121.566270       117.500000

After exploring our data, we can use our intuition, to drop some of the features that might have no contribution to our model

In [85]:
df = df.drop('number', axis=1)
df = df.drop('transaction_date', axis=1)
df.head()

house_age  distance_to_nearest_bus_station  number_of_convenience_stores  \
0       32.0                         84.87882                            10   
1       19.5                        306.59470                             9   
2       13.3                        561.98450                             5   
3       13.3                        561.98450                             5   
4        5.0                        390.56840                             5   

   latitude  longitude  unit_area_price  
0  24.98298  121.54024             37.9  
1  24.98034  121.53951             42.2  
2  24.98746  121.54391             47.3  
3  24.98746  121.54391             54.8  
4  24.97937  121.54245             43.1

### 4. Splitting the data into training and test sets

In [86]:
X = df.iloc[:, :-1]
y = df.iloc[:,df.shape[1]-1]

In [87]:
X.head()

house_age  distance_to_nearest_bus_station  number_of_convenience_stores  \
0       32.0                         84.87882                            10   
1       19.5                        306.59470                             9   
2       13.3                        561.98450                             5   
3       13.3                        561.98450                             5   
4        5.0                        390.56840                             5   

   latitude  longitude  
0  24.98298  121.54024  
1  24.98034  121.53951  
2  24.98746  121.54391  
3  24.98746  121.54391  
4  24.97937  121.54245

In [88]:
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.25,random_state=42)

### 5. Building the model to fit the training data

In [89]:
model = LinearRegression()
model.fit(X_train,y_train)

LinearRegression()

### 6. Using the model to make predictions for the test data set 

In [90]:
y_pred = model.predict(X_test)

In [91]:
print(f'b = {model.intercept_}')
print(f'w = {model.coef_}')

b = -1194.2145132626354
w = [-2.65600030e-01 -4.74964456e-03  1.13434674e+00  2.28565503e+02
 -3.67741638e+01]


### 7. Evaluating our model performance

In [92]:
print("Mean squared error: %.2f" % mean_squared_error(y_test, y_pred))
# The coefficient of determination: 1 is perfect prediction
print("Coefficient of determination: %.2f" % r2_score(y_test, y_pred))

Mean squared error: 66.75
Coefficient of determination: 0.58


We can use our model to make different kinds of predictions as follows

In [93]:
data = {
'house_age':[32.0],
'distance_to_nearest_bus_station': [84.87882],
'number_of_convenience_stores':[10],
'latitude':[24.98298],
'longitude':[121.54024],
}

data = pd.DataFrame(data)
val = model.predict(data)
print(val)

[48.93331686]


## Task 2
The aim of this task is to build a model which predicts sales based on the money spent on different platforms such as TV, radio, and newspaper for marketing by using Simple Linear Regression and Multiple Linear Regression. The dataset contains the following features ***TV, Radio, Newspaper*** and the target column is ***Sales*** 

Hint: _For Simple Linear Regression, select the feature with the highest correlation with the target column._

Compare the Mean Squared error and Coefficient of determination for both models

## Task 3:

To develop the best medical insurance products, the insurer needs access to historical data to approximate the medical costs of each user. With this data, a medical insurer can develop more accurate pricing models, plan a particular insurance outcome. For all these cases, the objective is to accurately predict insurance costs.

![alt text](img/health_insurance.jpg "Health Insurance")


#### Dataset
You are provided with a dataset in the **dataset/04-insurance_charges.csv** file that contains medical insurance records. The individual medical costs billed by health insurance are the target variable charges, and the rest of columns contain personal information such as age, gender, family status, and whether the patient smokes among other features.
The following is the description of the features contained in the dataset: 
- **age:** age of primary beneficiary
- **sex:** insurance contractor gender, female, male
- **bmi:** Body mass index, providing an understanding of body, weights that are relatively high or low relative to height,objective index of body weight (kg / m ^ 2) using the ratio of height to weight, ideally 18.5 to 24.9
- **children:** Number of children covered by health insurance / Number of dependents
- **smoker:** Smoking
- **region:** the beneficiary's residential area in the US, northeast, southeast, southwest, northwest.
- **charges:** Individual medical costs billed by health insurance

#### Use Case
The objective is to train a ML regression model that can accurately predict the insurance costs for a given individual

